lib/graphql/batch/execution_strategy.rb

```diff
@@ -1,14 +1,18 @@
 module GraphQL::Batch
   class ExecutionStrategy < GraphQL::Query::SerialExecution
     def execute(_, _, query)
-      Promise.sync(as_promise_unless_resolved(super))
+      deep_sync(super)
     rescue GraphQL::InvalidNullError => err
       err.parent_error? || query.context.errors.push(err)
       nil
     ensure
       GraphQL::Batch::Executor.current.clear
     end
 
+    def deep_sync(result)
+      Promise.sync(as_promise_unless_resolved(result))
```

Comment:
`as_promise_unless_resolved` recursively looks for promises in the result and wraps the whole result in a promise if it finds any.  Promise.sync only checks if the object itself is a promise